File: answers/19/main.cpp
#include <vector>
#include <chrono>
#include <iostream>
#include <opencv2/opencv.hpp>
#include <unistd.h>
#include <cmath>

void filter_LoG(const cv::Mat &src, cv::Mat &dst, cv::Size ksize=cv::Size(3, 3), float sigma=1.2) {
    dst = src.clone();
    dst.convertTo(dst, CV_32F);
    cv::Mat tmp = dst.clone();

    int w = src.size().width;
    int h = src.size().height;

    // prepare kernel
    cv::Mat k = cv::Mat::zeros(ksize, CV_32F);
    float tot_k = 0;
    for (int step_x=0; step_x<ksize.width; step_x++) {
        int x = step_x - (int)(ksize.width / 2);
        for (int step_y=0; step_y<ksize.height; step_y++) {
            int y = step_y - (int)(ksize.height / 2);
            k.at<float>(step_y, step_x) =
                (pow(x, 2) + pow(y, 2) - (2 * pow(sigma, 2)))
                / (2 * M_PI * pow(sigma, 6))
                * exp(-(pow(x, 2) + pow(y, 2))
                       / (2 * pow(sigma, 2)));
            tot_k += k.at<float>(step_y, step_x);
        }
    }
    k /= tot_k;

    dst.forEach<cv::Vec3f>([&](cv::Vec3f &pixel, const int pos[]) -> void {
        int x = pos[1];
        int y = pos[0];
        cv::Vec3f value = cv::Vec3f(0, 0, 0);
        for (int step_x=0; step_x<ksize.width; step_x++) {
            for (int step_y=0; step_y<ksize.height; step_y++) {
                int col = x + step_x - 1;
                int row = y + step_y - 1;
                if (col < 0) col = 0;
                if (col >= w) col = w - 1;
                if (row < 0) row = 0;
                if (row >= h) row = h - 1;
                value += tmp.at<cv::Vec3f>(row, col) * k.at<float>(step_y, step_x);
            }
        }
        pixel = value;
    });

    std::vector<cv::Mat> channels;
    cv::split(dst, channels);
    double mMin = INT_MAX;
    for (int ch=0; ch<3; ch++) {
        double min, max;
        cv::minMaxLoc(channels[ch], &min, &max);
        if (min < mMin) mMin = min;
    }
    double mMax = -INT_MAX;
    for (int ch=0; ch<3; ch++) {
        channels[ch] -= mMin;
        double min, max;
        cv::minMaxLoc(channels[ch], &min, &max);
        if (max > mMax) mMax = max;
    }
    for (int ch=0; ch<3; ch++) {
        channels[ch] /= mMax;
        channels[ch] *= 255;
    }
    cv::merge(channels, dst);
    dst.convertTo(dst, CV_8U);
}

int main() {
    cv::Mat img_orig = cv::imread("questions/dataset/images/imori_256x256_noise.png");
    cv::Mat img_result;
    filter_LoG(img_orig, img_result, cv::Size(5, 5), 3);
    cv::imshow("cpp_origin", img_orig);
    cv::imshow("cpp_result", img_result);
    cv::moveWindow("cpp_origin", 2*img_orig.size().width, 0);
    cv::moveWindow("cpp_result", 3*img_orig.size().width, 0);
    cv::waitKey(3000);
    return 0;
}

File: answers/19/main.py
import numpy as np
import cv2

def filter_LoG(img, k_size=(3, 3), sigma=1.2):
    # kernel
    k = np.zeros(k_size, dtype=np.float32)
    pad_x = k_size[0] // 2
    pad_y = k_size[1] // 2
    for x in range(-pad_y, -pad_y + k_size[1]):
        for y in range(-pad_x, -pad_x + k_size[0]):
            k[y + pad_y, x + pad_y] = (x ** 2 + y ** 2 - 2 * (sigma ** 2)) * np.exp( - (x ** 2 + y ** 2) / (2 * (sigma ** 2)))
    k /= k.sum()

    out = cv2.filter2D(img.astype(np.float32), -1, k)
    out -= out.min() # normalize > [0, 1]
    out /= out.max()
    return out

if __name__ == '__main__':
    img_orig = cv2.imread('questions/dataset/images/imori_256x256_noise.png')
    img_result = filter_LoG(img_orig, k_size=(5, 5), sigma=3)
    cv2.imshow('python_origin', img_orig)
    cv2.imshow('python_result', img_result)
    cv2.moveWindow('python_result', img_orig.shape[1], 0)
    cv2.waitKey(3000)
